editor/src/clj/internal/graph/types.clj

```diff
@@ -25,6 +27,49 @@
 (defn target-label [^Arc arc] (.target-label arc))
 (defn target [^Arc arc] [(.target-id arc) (.target-label arc)])
 
+(deftype Endpoint [^Long node-id ^Keyword label]
+  IHashEq
+  (hasheq [_]
+    (unchecked-add-int
+      (unchecked-multiply-int 31 (.hashCode node-id))
+      (.hasheq label)))
```

Comment:
We need to implement `IHashEq` to hit the fast path in `hash` used by Clojure's persistent data structures. We also use `.hasheq` on keywords because it's a simple field lookup (unlike `.hashCode`!)